src/main.cpp

```diff
@@ -2479,11 +2679,18 @@ bool DisconnectBlock(CBlock& block, CValidationState& state, CBlockIndex* pindex
                     LogPrint("cert", "%s():%d - SIDECHAIN-EVENT: failed cancelling scheduled event\n", __func__, __LINE__);
                     return error("DisconnectBlock(): ceasing height cannot be reverted: data inconsistent");
                 }
+                // Remove previously stored DataHash
+                if (!pblocktree->removeCertData(cert.GetScId(), cert.epochNumber))
```

Comment:
Seems, that `pblocktree` is not a proper storage, instead of using it we need to update state (CoinsDB) same as we did for SidechainInfo. And moreover, do the changes firstly to the cache, then do the `flush` for all changes at the same time.
To be discussed.